### ISSUE TITLE
[Bug]: PRIORITY scheduling can silently skip a running request for a full step when the preemption victim was already deferred earlier in the same schedule() call

### ISSUE THREAD

**Issue by ErenAta16** [vllm-project/vllm#49097]
### Your current environment

This is a pure Python control-flow bug in `vllm/v1/core/sched/scheduler.py`'s `Scheduler.schedule()`, in the `SchedulingPolicy.PRIORITY` preemption path. It does not depend on any hardware/kernel/CUDA specifics, so `collect_env.py` isn't meaningful here. Verified against `vllm-project/vllm` main at commit `e6d1310b2ac52e56e266450693af9de12a871a51` (2026-07-19), using the real `Scheduler`, `KVCacheManager`, `BlockPool`, and `Request` classes, no mocking of scheduler logic. The reproduction script below runs standalone (no pytest needed) in an environment with no GPU; the only stubbing involved is for platform-detection plumbing unrelated to the bug (noted inline in the script).

### 🐛 Describe the bug

**Summary:** under `SchedulingPolicy.PRIORITY`, when the scheduler needs to preempt a request to free KV blocks for the request currently being scheduled, and the chosen preemption victim happens to be a *different* running request that was already visited-and-skipped earlier in the *same* `schedule()` call (e.g. deferred by DP prefill-throttling, or otherwise given 0 new tokens this step), the scheduler's `req_index` bookkeeping is not adjusted for the removed list entry. This causes the very next request in `self.running` to be silently skipped for that entire scheduling step, even though it is `RUNNING`, was not preempted, has a pending token to compute, and there is available token budget and (after the preemption) a free KV block.

**Where**

`vllm/v1/core/sched/scheduler.py`, inside `Scheduler.schedule()`, the RUNNING-requests preemption loop (currently lines 570–603):

```python
                    # The request cannot be scheduled.
                    # Preempt the lowest-priority request.
                    if self.policy == SchedulingPolicy.PRIORITY:
                        preempted_req = max(
                            self.running,
                            key=lambda r: (r.priority, r.arrival_time),
                        )
                        self.running.remove(preempted_req)
                        if preempted_req in scheduled_running_reqs:
                            preempted_req_id = preempted_req.request_id
                            scheduled_running_reqs.remove(preempted_req)
                            token_budget += num_scheduled_tokens.pop(preempted_req_id)
                            req_to_new_blocks.pop(preempted_req_id)
                            scheduled_spec_decode_tokens.pop(preempted_req_id, None)
                            preempted_encoder_inputs = scheduled_encoder_inputs.pop(
                                preempted_req_id, None
                            )
                            if preempted_encoder_inputs:
                                # Restore encoder compute budget if the preempted
                                # request had encoder inputs scheduled in this step.
                                num_embeds_to_restore = sum(
                                    preempted_req.get_num_encoder_embeds(i)
                                    for i in preempted_encoder_inputs
                                )
                                encoder_compute_budget += num_embeds_to_restore
                            req_index -= 1
                    else:
                        preempted_req = self.running.pop()

                    self._preempt_request(preempted_req, scheduled_timestamp)
                    preempted_reqs.append(preempted_req)
                    if preempted_req == request:
                        # No more request to preempt. Cannot schedule this request.
                        break
```

`preempted_req` is chosen via `max(self.running, key=lambda r: (r.priority, r.arrival_time))` — the globally worst-priority *running* request, regardless of its position in `self.running` (which is plain arrival/insertion order; it is never sorted by priority). `req_index` is only decremented when the victim was already in `scheduled_running_reqs` (i.e. it was fully scheduled earlier this same round). There is no corresponding handling for the case where the victim is at an index `< req_index` but was **not** scheduled this round — i.e. it was already visited earlier in this same `schedule()` call and skipped via one of the `continue` paths (deferred prefill under DP throttling, pipeline-parallel eligibility gating, encoder-budget exhaustion, etc. — any path that leaves a request in `self.running` without adding it to `scheduled_running_reqs`).

When that happens, `self.running.remove(preempted_req)` shifts every later element one slot left, but `req_index` is left unchanged. On the next outer-loop iteration, `self.running[req_index]` now points one position past where it should, so the request that should be there is never visited this step.

For contrast, `vllm/v1/core/sched/scheduler.py` also has a *different*, already-known issue in the same neighborhood: PR #36390 (open) fixes the case where the preemption cascade fails entirely and `preempted_req == request` (self-preemption), which used to `break` the whole loop. That fix is orthogonal to this bug: in the scenario below the cascade *succeeds* (the retry after preempting the victim gets its blocks), so `new_blocks is not None` and the code path #36390 touches is never reached. This is a distinct bug in the same function.

### Reproduction

This is a real, deterministic reproduction against the actual scheduler code (no mocking of `Scheduler`, `KVCacheManager`, `BlockPool`, or `Request`; only unrelated platform/import plumbing is stubbed, commented inline). Save as `repro.py` and run with `python repro.py`:

```python
"""
Reproduction of a request-skipping bug in the PRIORITY-policy preemption path
of vllm/v1/core/sched/scheduler.py :: Scheduler.schedule().
"""
import os
import sys
import types

VLLM_ROOT = "/path/to/vllm"  # <- point this at your vllm checkout

os.environ["HF_HUB_OFFLINE"] = "1"
os.environ["VLLM_LOGGING_LEVEL"] = "WARNING"
os.environ["PYTHONPATH"] = VLLM_ROOT + os.pathsep + os.environ.get("PYTHONPATH", "")
sys.path.insert(0, VLLM_ROOT)

# --- The following stubs are unrelated to the bug: they only work around
# --- this being a non-GPU / non-pip-installed checkout. On a normal vllm
# --- dev install (e.g. `pip install -e .`) none of this is needed -- just
# --- use the real `create_scheduler_with_priority` / `create_requests_with_priority`
# --- helpers from tests/v1/core/test_scheduler.py directly.
if sys.platform == "win32":
    sys.modules["uvloop"] = types.ModuleType("uvloop")
    sys.modules["uvloop"].run = lambda *a, **k: None
try:
    import xgrammar  # noqa: F401
except ImportError:
    xg = types.ModuleType("xgrammar")
    xg.GrammarMatcher = type("GrammarMatcher", (), {})
    xg.CompiledGrammar = type("CompiledGrammar", (), {})
    sys.modules["xgrammar"] = xg

import torch
from vllm.config import CacheConfig, ModelConfig, ParallelConfig, SchedulerConfig, VllmConfig
from vllm.sampling_params import SamplingParams
from vllm.utils.hashing import sha256
from vllm.v1.core.kv_cache_utils import get_request_block_hasher, init_none_hash
from vllm.v1.core.sched.scheduler import Scheduler
from vllm.v1.core.single_type_kv_cache_manager import register_all_kvcache_specs
from vllm.v1.kv_cache_interface import FullAttentionSpec, KVCacheConfig, KVCacheGroupSpec
from vllm.v1.outputs import ModelRunnerOutput
from vllm.v1.request import Request, RequestStatus
from vllm.v1.structured_output import StructuredOutputManager

try:
    import vllm.platforms
    from vllm.platforms.cpu import CpuPlatform
    vllm.platforms.current_platform = CpuPlatform()
    import vllm.platforms.cpu as _cpu_mod
    _cpu_mod.get_max_threads = lambda pid=0: os.cpu_count() or 1
except Exception:
    pass  # not needed on a normal Linux/macOS dev install

BLOCK_SIZE = 16
NUM_BLOCKS = 6  # 1 null block -> 5 usable
MAX_MODEL_LEN = 1000
LONG_PREFILL_THRESHOLD = 15
EOS_TOKEN_ID = 50256


def build_scheduler():
    model_config = ModelConfig(
        model="facebook/opt-350m", trust_remote_code=True, dtype="float16",
        seed=42, skip_tokenizer_init=True, max_model_len=MAX_MODEL_LEN,
    )
    scheduler_config = SchedulerConfig(
        max_num_seqs=5, max_num_batched_tokens=200, max_model_len=MAX_MODEL_LEN,
        long_prefill_token_threshold=LONG_PREFILL_THRESHOLD,
        enable_chunked_prefill=True, is_encoder_decoder=model_config.is_encoder_decoder,
        policy="priority", watermark=0.0,
        scheduler_reserve_full_isl=False,  # let chunked prefill actually chunk
    )
    cache_config = CacheConfig(block_size=BLOCK_SIZE, gpu_memory_utilization=0.9,
                                cache_dtype="auto", enable_prefix_caching=False)
    vllm_config = VllmConfig(scheduler_config=scheduler_config, model_config=model_config,
                              cache_config=cache_config, parallel_config=ParallelConfig())
    kv_cache_spec = FullAttentionSpec(block_size=BLOCK_SIZE, num_kv_heads=1, head_size=1, dtype=torch.float32)
    kv_cache_config = KVCacheConfig(num_blocks=NUM_BLOCKS, kv_cache_tensors=[],
                                     kv_cache_groups=[KVCacheGroupSpec(["layer"], kv_cache_spec)])
    cache_config.num_gpu_blocks = NUM_BLOCKS
    register_all_kvcache_specs(vllm_config)
    return Scheduler(vllm_config=vllm_config, kv_cache_config=kv_cache_config, block_size=BLOCK_SIZE,
                      log_stats=True, structured_output_manager=StructuredOutputManager(vllm_config))


_hash_init = False


def make_request(req_id, num_tokens, priority, arrival_time):
    global _hash_init
    if not _hash_init:
        init_none_hash(sha256)
        _hash_init = True
    block_hasher = get_request_block_hasher(BLOCK_SIZE, sha256)
    sp = SamplingParams(ignore_eos=True, max_tokens=50)
    sp.update_from_generation_config({}, EOS_TOKEN_ID)
    return Request(request_id=req_id, prompt_token_ids=[7] * num_tokens, sampling_params=sp,
                    pooling_params=None, priority=priority, arrival_time=arrival_time,
                    block_hasher=block_hasher)


def mock_output(out, tok=100):
    ids = list(out.num_scheduled_tokens.keys())
    return ModelRunnerOutput(req_ids=ids, req_id_to_index={r: i for i, r in enumerate(ids)},
                              sampled_token_ids=[[tok] for _ in ids], logprobs=None,
                              prompt_logprobs_dict={}, pooler_output=[])


def main():
    scheduler = build_scheduler()

    # R1: admit A -- long prompt, WORST priority. Chunked prefill (15 tok/round).
    A = make_request("A", num_tokens=200, priority=9, arrival_time=1.0)
    scheduler.add_request(A)
    out = scheduler.schedule()
    scheduler.update_from_output(out, mock_output(out))

    # R2: admit B -- short prompt (15 tok, fits in 1 shot), BEST priority.
    # B moves straight to decode phase (is_prefill_chunk=False) this round.
    B = make_request("B", num_tokens=LONG_PREFILL_THRESHOLD, priority=0, arrival_time=2.0)
    scheduler.add_request(B)
    out = scheduler.schedule()
    scheduler.update_from_output(out, mock_output(out))

    # R3: admit C -- 1-token prompt, middle priority. B decodes 1 tok (15->16,
    # exactly fills its first block).
    C = make_request("C", num_tokens=1, priority=1, arrival_time=3.0)
    scheduler.add_request(C)
    out = scheduler.schedule()
    scheduler.update_from_output(out, mock_output(out))

    free = scheduler.kv_cache_manager.block_pool.get_num_free_blocks()
    print(f"[R3] free blocks = {free} (expect 0: A=3 blocks + B=1 + C=1 == 5 usable)")
    assert free == 0
    print(f"[R4] scheduler.running = {[r.request_id for r in scheduler.running]}")

    # R4 (TARGET STEP): B needs a 2nd block (computed 16->17) but 0 free.
    # PRIORITY preemption picks A (worst priority) -- but A was already
    # skipped THIS round (defer_prefills deferred its still-in-progress
    # chunk), so it's not in scheduled_running_reqs, and req_index is not
    # decremented when it's removed from self.running.
    out = scheduler.schedule(throttle_prefills=True)

    scheduled = sorted(out.num_scheduled_tokens.keys())
    print(f"[R4] scheduled this step = {scheduled}")
    print(f"[R4] A.status={A.status}  B.status={B.status}  C.status={C.status}")
    print(f"[R4] free blocks after = {scheduler.kv_cache_manager.block_pool.get_num_free_blocks()}")

    if "C" not in out.num_scheduled_tokens and C.status == RequestStatus.RUNNING:
        print("BUG CONFIRMED: C is RUNNING, not preempted, had a pending "
              "decode token and a free block, yet got 0 scheduled tokens.")
        out2 = scheduler.schedule()
        print(f"[R5] scheduled next step = {sorted(out2.num_scheduled_tokens.keys())} "
              "(C resumes normally -- confirms this is a dropped step, not a deadlock)")
        sys.exit(1)
    else:
        print("Not reproduced.")
        sys.exit(0)


if __name__ == "__main__":
    main()
```

**Actual output when run against the real, unmodified `Scheduler`:**

```
[R3] free blocks = 0 (expect 0: A=3 blocks + B=1 + C=1 == 5 usable)
[R4] scheduler.running = ['A', 'B', 'C']
[R4] scheduled this step = ['B']
[R4] A.status=RequestStatus.PREEMPTED  B.status=RequestStatus.RUNNING  C.status=RequestStatus.RUNNING
[R4] free blocks after = 2
BUG CONFIRMED: C is RUNNING, not preempted, had a pending decode token and a free block, yet got 0 scheduled tokens.
[R5] scheduled next step = ['A', 'C'] (C resumes normally -- confirms this is a dropped step, not a deadlock)
```

Step by step at R4: `self.running = [A, B, C]`, `req_index=0` visits `A`, which `defer_prefills` skips via `continue` (still mid-prefill, deferred this DP-throttled step) — `req_index` becomes 1, `A` stays in `self.running`, not in `scheduled_running_reqs`. `req_index=1` visits `B`, which needs a new block that isn't available; the preemption cascade picks `A` (worst priority) via `max()`, removes it from `self.running` (now `[B, C]`), does not decrement `req_index` (since `A` isn't in `scheduled_running_reqs`), frees `A`'s blocks, and the retry for `B` succeeds. `B` is scheduled and `req_index` becomes 2. The outer loop condition `req_index < len(self.running)` is now `2 < 2`, so the loop exits — `C` is never visited this step.

### Expected behavior

`C` should be considered for scheduling in this step: it is `RUNNING`, was not the preemption target, needs only 1 token that fits in its already-allocated block, and there is a free block and ample token budget available (including the block just freed from `A`'s preemption).

### Actual behavior

`C` receives 0 scheduled tokens for this step with no state change (not preempted, not finished, not blocked on anything) — it is silently dropped purely because of the `req_index` bookkeeping gap. It resumes normally on the very next `schedule()` call, so this is not a deadlock, but every time this shape recurs under sustained KV pressure with `SchedulingPolicy.PRIORITY`, some otherwise-schedulable running request loses a full decoding step for no observable reason (no log line, no status change), which shows up as reduced throughput / an unfair scheduling gap that's hard to attribute to anything from the outside.

### Suggested fix

In the same `if preempted_req in scheduled_running_reqs:` block (`vllm/v1/core/sched/scheduler.py`, currently around line 578), add an `elif` (or restructure the condition) that also decrements `req_index` whenever the victim's original position in `self.running` was `< req_index`, regardless of whether it was in `scheduled_running_reqs` — e.g. track the victim's index via `self.running.index(preempted_req)` (or a position lookup) before removing it, and decrement `req_index` whenever that index is `< req_index`. The current code only special-cases the "already fully scheduled this round" case; the "already visited but skipped this round" case needs the same `req_index -= 1` treatment since both remove an element at a position strictly before the current cursor.

### Are you a ML Ops Team?

No

### What vLLM version are you on?

`main` branch, verified against commit `e6d1310b2ac52e56e266450693af9de12a871a51` (2026-07-19). The relevant code (`Scheduler.schedule()`'s PRIORITY preemption block) has been present in this form across recent history; I did not bisect further back since the bug is a logic gap rather than a regression from a specific commit.

**Comment by mrcha033** [comment]
It seems the priority branch decrements req_index only when the victim is in scheduled_running_reqs; it must also account for any removed victim whose list position precedes the current cursor.

**Comment by ErenAta16** [comment]
That's the right read of it. To be precise about the boundary: it only matters when the victim's index in self.running is less than req_index at the time of removal, since anything at or after req_index hasn't been visited yet this step and removing it just shifts the not-yet-visited tail left, which the outer loop's own indexing already tolerates. The case that actually breaks is a victim already visited-and-skipped earlier in this same schedule() call, sitting behind req_index.
